{
 "cells": [
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **S**ubstitution **I**nferences  using **M**athematical **B**iology **A**pproach (**SIMBA**) \n",
    "\n",
    "### _by Milton Simbarashe Kambarami_\n",
    "\n",
    "_in fulfilment of Master of Philosophy (MPhil) in Bioinformatics under Faculty of Medicine and Health Sciences, University of Zimbabwe._\n",
    "\n",
    "Substitution Inference using Mathematical Biology Approach (SIMBA) is a package which uses a mathematical approach based on Nei and Gojobori, 1986 to calculate numbers of synonynmous and non-synonymous substitution and further modified to calculate selection pressure by comparing values of synonymous and non-synonymous substitution obtained.\n",
    "\n",
    "It is worthnoting that different approaches usually reach close values especially when working with similar codons. The SIMBA approach like most evolutionary analysis approaches treats each codon as an independent unit undergoing selection without effects of neighbouring codons or homologous codons in other sequences. When working with too divergent sequences the obtained values might be biased, however because this is an approach to be useful when working with Spike genes of SARS CoV 2, the bias is minimised. \n",
    "\n",
    "My motivation for developing this package is the unavailability of __specific__ tools to analyse the Spike gene of Coronaviruses and the computational expensiveness of available tools when trying to generalise analysis of all sequences, convergent or divergent. SIMBA is specific for the spike gene analysis and uses a very fast lightweight mathematical approach to find substitution inferences. However, analysis of other genes can be done by tweaking certain parts of the code which the author can freely do upon request.\n",
    "\n",
    "NB for the sake of shortness of code, I used the word codon as short for triplet code in this package.\n",
    "\n",
    "A number of approaches used to ensure faster inferences include:\n",
    "1. Use a fast Nei and Gojobori, 1986 mathematical approach\n",
    "2. Assumption codons are independent evolutionary units.\n",
    "3. Exclusion of weighting since its going to be used on very similar amino acids."
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Spike Gene Cutter\n",
    "\n",
    "So far there is no simple process of cutting and cleaning Spike gene sequences from SARS-CoV-2 whole genome sequences. However, the user is supposed to use alignment tool of choice after the spike gene has been extracted. User can cut their genes using this block of code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def spike_gene_cutter_cleaner(input_file, file_format):\n",
    "\n",
    "    # from Bio import SeqIO    # We use Biopython's SeqIO parser to load our sequences\n",
    "    import sys\n",
    "\n",
    "    # The  next section loads the SARS-CoV-2 sequences from file.\n",
    "\n",
    "    input_file = input(\"Enter full 'input' filename and filetype eg input_file.fasta\")\n",
    "    file_format = input(\"Enter 'file format' eg fasta\")\n",
    "\n",
    "    # Extraction of the approximate S gene\n",
    "    spike_genes = []        # All extracted spike genes will be stored in this list\n",
    "\n",
    "    for seq_record in SeqIO.parse(input_file, file_format):\n",
    "        spike_cut = seq_record[21500:25500]\n",
    "        spike_genes.append(spike_cut)\n",
    "    \n",
    "\n",
    "    spike_file = SeqIO.write(spike_genes, 'Spike_gene_file.fasta','fasta')\n",
    "\n",
    "    # Next part removes spike genes with a number of Unidentified nucleotides more than 2\n",
    "\n",
    "    clean_spike_genes = [] # This list will store cleaned spike genes\n",
    "\n",
    "    for spike_record in SeqIO.parse('Spike_gene_file.fasta','fasta'):\n",
    "        # first its wiser to use a string object of the sequence for counting additionally all letters should be ensured to be in upper case\n",
    "\n",
    "        string_spike = (str(spike_record.seq)).upper()\n",
    "\n",
    "        if string_spike.count('N') > 2:              #Unidentified nucleotides more than 2\n",
    "            if spike_record.seq not in clean_spike_genes:  # Removing repeat sequences\n",
    "                clean_spike_genes.append(spike_record)      #spike record is appended instead of string_spike so it does not lose Sequence object attributes\n",
    "\n",
    "\n",
    "print('Please check current folder for the Spike gene sequences which were extracted')\n",
    "\n",
    "\n",
    "# In case the user prefers to see extracted sequences\n",
    "print_sample = input('Do you want to view snips of extracted sequences: Y/N ')\n",
    "\n",
    "if print_sample.upper() == 'Y':\n",
    "    for spike in SeqIO.parse('Spike_gene_file.fasta','fasta'):\n",
    "            print(f'Length of approximate spike_gene = {len(spike)}')\n",
    "            print(f'Representative spike sequence = {repr(spike.seq)}')\n",
    "            print(f'Sequence id = {spike.id}')\n",
    "\n",
    "elif print_sample.upper() == 'N':\n",
    "        pass\n",
    "\n",
    "else:\n",
    "    print(\"Please enter 'Y' for 'Yes' and 'N' for 'No'\")\n",
    "\n",
    "# User can use alignment tool of choice and it is recommended to align against the GISAID Ref Seq for more accurate\n",
    "# downstream processes"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## The Triplet Code Table\n",
    "\n",
    "Synonymous substitutions maybe used as a molecular clock for dating the evolutionary time of closely related species.\n",
    " \n",
    " <p align ='center'>\n",
    " <img\n",
    "    src = 'DNA triplet code.png'\n",
    ">\n",
    "</p>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import AlignIO\n",
    "\n",
    "aligned_sequences_filename = input('Enter filename including format eg aligned_spike.fasta: ')     #make sure the alignment file is in in the same\n",
    "                                                                                                #directory as SIMBA else use full filepath\n",
    "aligned_sequences_format = input('Enter file format eg fasta: ')\n",
    "\n",
    "aligned_sequences = AlignIO.read(aligned_sequences_filename, aligned_sequences_format)\n",
    "\n",
    "aligned_sequences = ()\n",
    "\n",
    "def translator(codon):\n",
    "\n",
    "    '''Triplet codes, (codon for simplicity) that code for specific amino acids based on the DNA triplet code table'''\n",
    "\n",
    "    for codon in aligned_sequences:\n",
    "        # Gene sequence is a fasta file from a codon-aligned sequences, loaded using AlignIO of Biopython.\n",
    "        # Reminder that all loaded DNA sequences must be in upper case and in string format\n",
    "        if codon == 'TTT' or 'TTC':\n",
    "            amino_acid = 'F'\n",
    "\n",
    "        elif codon == 'TTA' or 'TTG' or 'CTT' or 'CTC'or 'CTA' or 'CTG':\n",
    "            amino_acid = 'L'\n",
    "\n",
    "        elif codon == 'ATT' or 'ATC' or 'ATA':\n",
    "            amino_acid = 'I'\n",
    "\n",
    "        elif codon == 'ATG':\n",
    "            amino_acid = 'M'\n",
    "\n",
    "        elif codon == 'GTT' or 'GTC' or 'GTA' or 'GTG':\n",
    "            amino_acid = 'V'\n",
    "\n",
    "        elif codon == 'TCT' or 'TCC' or 'TCA' or 'TCG' or 'AGT' or 'AGC':\n",
    "            amino_acid = 'S'\n",
    "\n",
    "        elif codon == 'CCT' or 'CCC' or 'CCA' or 'CCG':\n",
    "            amino_acid = 'P'\n",
    "\n",
    "        elif codon == 'ACT' or 'ACC' or 'ACA' or 'ACG':\n",
    "            amino_acid = 'T'\n",
    "\n",
    "        elif codon == 'GCT' or 'GCC' or 'GCA' or 'GCG':\n",
    "            amino_acid = 'A'\n",
    "\n",
    "        elif codon == 'TAT' or 'TAC':\n",
    "            amino_acid = 'Y'\n",
    "        \n",
    "        elif codon == 'TAA' or 'TAG' or 'TGA':\n",
    "            amino_acid = '*'\n",
    "\n",
    "        elif codon == 'CAT' or 'CAC':\n",
    "            amino_acid = 'H'\n",
    "\n",
    "        elif codon == 'CAA' or 'CAG':\n",
    "            amino_acid = 'Q'\n",
    "\n",
    "        elif codon == 'AAA' or 'AAG':\n",
    "            amino_acid = 'K'\n",
    "\n",
    "        elif codon == 'GAA' or 'GAG':\n",
    "            amino_acid = 'E'\n",
    "\n",
    "        elif codon == 'TGT' or 'TGC':\n",
    "            amino_acid = 'C'\n",
    "\n",
    "        elif codon == 'TGG':\n",
    "            amino_acid = 'W'\n",
    "\n",
    "        elif codon == 'CGT' or 'CGC' or 'CGA' or 'CGG' or 'AGA' or 'AGG':\n",
    "            amino_acid = 'R'\n",
    "\n",
    "        elif codon == 'GGT' or 'GGC' or 'GGA' or 'GGG':\n",
    "            amino_acid = 'G'\n",
    "        \n",
    "        elif 'N' IS IN codon:\n",
    "            amino_acid = '?'\n",
    "\n",
    "        else:\n",
    "            print('Please check if all bases are A, C, T or G')"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sum of synonymous substitution in a codon at each _i_ th site\n",
    "\n",
    "Genetic code table indicates that all substitutions at the second nucleotide positions of codons result in amino acid whereas a fraction of the nucleotide changes at the first and third positions are synonymous.\n",
    "\n",
    "Under the assumption of equal nucleoties frequencies and random substitution, this fraction is ~5 % for the first position and ~72 % for the thrid position.\n",
    "\n",
    "> $f_i$ = fraction of synonymous changes at the _i_ th position of a given codon (i = 1,2,3)\n",
    "\n",
    "> _s_ = sum of synonymous substitution at each site\n",
    "\n",
    "> _n_ = sum of non-synonymous substitution at each site\n",
    "\n",
    "The _n_ and _s_ for this codon are then given by:\n",
    "\n",
    "$$\n",
    "\\begin{align}\n",
    "s = \\sum_{i=1}^{3} f_i \\  \\text{\\ and\\  n=(3-s)}\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "> using __Leu__ as an example,\n",
    "\n",
    "$f_1 = \\frac{1}{3}  \\text{A} \\to \\text{G}$\n",
    "\n",
    "using genetic code table, there is 1 in 3 chances that a change is from T $\\to$ C.\n",
    "\n",
    "$f_2$ = 0,   \n",
    "\n",
    "$f_3 =  \\frac{1}{3}(A \\to G)$ thus, \n",
    "\n",
    "$ S = \\frac{1}{3} + 0 + \\frac{1}{3} = \\frac{2}{3}$\n",
    "\n",
    "$ n = \\frac{2}{3} + \\frac{3}{3} + \\frac{2}{3} = \\frac{7}{3}$\n",
    "\n",
    "Approach of method:\n",
    "\n",
    "1. Load aligned sequences.\n",
    "2. Function which assigns each codon a codon_site number (psi_j) in each sequence\n",
    "3. Compare each base at _i_ th position for each codon at $\\psi_j$ where $\\psi_j$ = codon at site `j` by comparing all \n",
    "   codons at $\\psi_j$ with $\\psi_j$ = 0\n",
    "4. Find sum of synonymous substitution fractions (s) at each $\\psi_j$ and append to a sum of synonymous sites list\n",
    "5. Find sum of non-synonymous substitution fraction (n) at each $psi_j$.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#1. Load aligned sequences\n",
    "\n",
    "from Bio import AlignIO\n",
    "\n",
    "aligned_sequences_filename = input('Enter filename including format eg aligned_spike.fasta: ')     #make sure the alignment file is in in the same\n",
    "                                                                                                #directory as SIMBA else use full filepath\n",
    "aligned_sequences_format = input('Enter file format eg fasta: ')\n",
    "\n",
    "aligned_sequences = AlignIO.read(aligned_sequences_filename, aligned_sequences_format)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# 2. Function which assigns each codon a codon_site number (psi_j) in each sequence\n",
    "\n",
    "def codon_splitter(aligned_sequences):\n",
    "    codoned_sequences = []                     #this list holds all sequences but in their codon form, each sequence is a list of codons\n",
    "    for seq in aligned_sequences:\n",
    "        codon_list = []                        #this list holds each sequence as an array of codons.\n",
    "        for i in range(0, len(seq), 3):\n",
    "            codon_list.append(seq[i:i+3])\n",
    "        return codon_list\n",
    "    codoned_sequences.append(codon_list)\n",
    "\n",
    "    # the next part seperates lists of sequences in codon form inside the codon_sequences list\n",
    "    for index,item in enumerate(codoned_sequences):\n",
    "        codoned_sequences[index] = codoned_sequences[index][0].split(\",\")\n",
    "        return codoned_sequences\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#3. Compare each base at _i_ th position for each codon at codon_j where codon_j = codon at site `j`\n",
    "#4. Find sum of synonymous substitution fraction and append to a sum of codon synonymous substitution fractions list\n",
    "\n",
    "\n",
    "list_of_codon_synonymous_substitutions_fractions = []      #list to collect sum of codon synonymous substitutions_fractions  per site\n",
    "                                                           #code must be added which k\n",
    "\n",
    "def sum_of_codon_synonymous_substitutions(codoned_sequences, lower=0, upper=2):\n",
    "    psi_j = [ys + [x] for x, ys in zip(codoned_sequences)]       #this code selects items(i.e. codons) with the same index in codoned sequences list lists \n",
    "\n",
    "    for psi_j in codoned_sequences:\n",
    "        ref_codon = psi_j[0]\n",
    "\n",
    "        for i,nucleotide in enumerate(zip(psi_j, ref_codon)):      #Lower represents the lower bound in our sum and upper is the upper bound. For python index, the first value would be0\n",
    "                                                    #i is the index of a base in a given codon_site\n",
    "            if i == 0:                      \n",
    "                if nucleotide != ref_codon:            \n",
    "                    fraction_of_change1 = 0.05            #Fraction of change denotes the chances of a codon for coding another amino acid when there is a mutation eg \n",
    "                else:\n",
    "                    fraction_of_change1= 0                                    #changing the first codon nucleotide results in 5% chance of change in amino acid coded.                     \n",
    "            \n",
    "            elif i == 1:\n",
    "                if nucleotide != ref_codon:\n",
    "                    fraction_of_change2 = 1\n",
    "                else:\n",
    "                    fraction_of_change2 = 0\n",
    "\n",
    "            elif i == 2:\n",
    "                if nucleotide != ref_codon:\n",
    "                    fraction_of_change3 = 0.72\n",
    "                else:\n",
    "                    fraction_of_change3 = 0\n",
    "\n",
    "            sum_of_synonymous_fractions = sum(fraction_of_change1 + fraction_of_change2 + fraction_of_change3)              #gives use the sum of fraction of synonymous changes at ith position.\n",
    "            list_of_codon_synonymous_substitutions_fractions.append(sum_of_synonymous_fractions)\n",
    "            \n",
    "            return sum_of_synonymous_fractions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#5. Find sum of non-synonymous substitution fraction (n) at each $psi_j$\n",
    "\n",
    "def sum_of_non_synonymous_codon_substitutions(sum_of_synonymous_fractions):\n",
    "    sum_of_non_synonymous_fractions = 3 - sum_of_synonymous_fractions   # to find sum of non-synonymous fraction we subtract the sum of \n",
    "\n",
    "    return sum_of_non_synonymous_fractions                               #synonymous fraction from 3\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mean of synonymous sites (S) and non_synonymous sites (N)\n",
    "\n",
    "For a DNA sequence of __r__ codons, the total number of synonymous and non-synonymous sites at $\\psi_j$ is therefore given by:\n",
    "\n",
    "$$ \n",
    "\\begin{align*}\n",
    "S = \\sum_{j=1}^r S_i \\     \\text{and\\   N = (3r -S)} \\\\\n",
    "\\text{where} \\  s_i = \\text{value of s for the i-th codon} \\\\\n",
    "\\text{j = position number of codon j in DNA sequence with r codons}\n",
    "\\end{align*}\n",
    "$$\n",
    "\n",
    "\n",
    "When two sequences are compared, the averages of __S__ and __N__ are used.\n",
    "\n",
    "\n",
    "Therefore:\n",
    "\n",
    "$$\n",
    "\\begin{align*}\n",
    "S = \\sigma \\text{ (list of codon synonymous substitutions fractions)}  \\\\\n",
    "\\text{where}  \\    \\sigma = \\text{mean/average at } \\psi_j \\text{ obtained in s}  \\\\\n",
    "S = \\frac{\\text{sum of elements in list of codon synonymous substitutions fractions}}{\\text{number of elements in list of codon synonymous substitutions fractions }}  \\\\\n",
    "\\end{align*}\n",
    "$$"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statistics\n",
    "\n",
    "# mean_number_synonymous_substitution_list = []\n",
    "# mean_number_non_synonymous_substitution_list = []\n",
    "#these lists will be converted to dict to keep their `j` and added into the dataframe for visualisation.\n",
    "\n",
    "\n",
    "def mean_number_synonymous_substitution_at_psi_j(list_of_codon_synonymous_substitution_fractions):\n",
    "    mean_number_synonymous_substitution = int(statistics.mean((list_of_codon_synonymous_substitutions_fractions)))\n",
    "    # mean_number_synonymous_substitution_list.append(mean_number_synonymous_substitution)\n",
    "\n",
    "def mean_number_non_synonymous_substitution_at_psi_j(list_of_codon_synonymous_substitutions_fractions,mean_number_synonymous_substitution):\n",
    "    r = len(list_of_codon_synonymous_substitutions_fractions)\n",
    "    mean_number_non_synonymous_substitution = ((3*r) - (mean_number_synonymous_substitution))\n",
    "    # mean_number_non_synonymous_substitution_list.append(mean_number_non_synonymous_substitution)"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing nucleotide differences between a pair of homologous sequences\n",
    "\n",
    "To compute the number of synonymous and non synonymous nucleotide differences between a pair of homologous sequences, we compare the two sequences codon, we compare the two sequences codon by codon and count the number of synoymous and non-synonymous nucleotide differences  for each pair of codon compared.\n",
    "\n",
    "When there is only one nucleotide difference, we can immediately decide whether the substitution is synonymous or non-synonymous.\n",
    "> eg, if the codon pairs compared are __GTT__ (_Val_) and __GTA__ (_Val_), there is one synonymous difference.\n",
    "\n",
    "We denote $S_d$ and $n_d$ the number of synonymous and non-synoynymous differences per codon, respecitively. In the prent case, $S_d = 1$ and $n_d = 0$.\n",
    "\n",
    "For example, in the comparison of __TTT__ and __GTA__, the two pathways are as follows:\n",
    "\n",
    "> _Pathway I_:\n",
    "> > __TTT__ (_Phe_) $\\to$ __GTT__ (_Val_) $\\to$ __GTA__ (_Val_)\n",
    "\n",
    "> _Pathway II_:\n",
    "> > __TTT__ (_Phe_) $\\to$ __TTA__ (_Leu_) $\\to$ __GTA__ (_Val_)\n",
    "\n",
    "Pathway I involves one synonymous and one non-synonymous substitution whereas Pathway II involves two non-synonymous  substitutions.\n",
    "\n",
    "We assume that pathway I and II occur with equal probability.\n",
    "\n",
    "The $s_d$ and $n_d$ then become 0.5 and 1.5 repsectively.\n",
    "\n",
    "When there are three nucleotide differences between the codons compared, there are six different possible pathwaysbetween the codons [3(x-1)] and in each pathway there are 3 mutation steps [x] where x = nucleotide differences.\n",
    "\n",
    "It is now clear that the total number of synonymous and non-synonymous differences can be obtained by summing up these values over all codons i.e.\n",
    "$$\n",
    "\\begin{align*}\n",
    "S_d = \\sum_{j=1}^{r} s_{dj} \\\\\n",
    "N_d = \\sum_{j=1}^{r} n_{dj} \\\\\n",
    "\\text{where} \\ s_{dj} \\text{\\  and } n_{dj} \\text{\\  are } s_d \\text{\\ and } n_d \\text{ for the j-th codon respectively} \\\\\n",
    "\\text{and r is the number of codons compared}\n",
    "\\end{align*}\n",
    "$$\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import statistics\n",
    "\n",
    "synonymous_differences_list = []\n",
    "non_synonymous_differences_list = []\n",
    "\n",
    "def selection_differences_number(psi_j):\n",
    "    psi_j = [ys + [x] for x, ys in zip(codoned_sequences)]       #this code selects items(i.e. codons) with the same index in codoned sequences list lists \n",
    "\n",
    "    for psi_j in codoned_sequences:\n",
    "        ref_codon = psi_j[0]\n",
    "\n",
    "        for i,nucleotide in enumerate(zip(psi_j, ref_codon)): \n",
    "            if i == 0:\n",
    "                if nucleotide == ref_codon:\n",
    "                    synonymous_difference = 0\n",
    "                    non_synonymous_difference = 0\n",
    "\n",
    "                elif nucleotide != ref_codon:\n",
    "                    if translator(nucleotide) == translator(ref_codon):\n",
    "                        synonymous_difference = 1\n",
    "                        synonymous_differences_list.append(synonymous_difference)\n",
    "                    \n",
    "                    elif translator(nucleotide) != translator(ref_codon):\n",
    "                        non_synonymous_difference = 1\n",
    "                        non_synonymous_differences_list.append(non_synonymous_difference)\n",
    "            \n",
    "            if i == 1:\n",
    "                if nucleotide == ref_codon:\n",
    "                    synonymous_difference = 0\n",
    "                    non_synonymous_difference = 0\n",
    "\n",
    "                elif nucleotide != ref_codon:\n",
    "                    if translator(nucleotide) == translator(ref_codon):\n",
    "                        synonymous_difference = 1\n",
    "                        synonymous_differences_list.append(synonymous_difference)\n",
    "                    \n",
    "                    elif translator(nucleotide) != translator(ref_codon):\n",
    "                        non_synonymous_difference = 1\n",
    "                        non_synonymous_differences_list.append(non_synonymous_difference)\n",
    "\n",
    "            elif i == 2:\n",
    "                if nucleotide == ref_codon:\n",
    "                    synonymous_difference = 0\n",
    "                    non_synonymous_difference = 0\n",
    "\n",
    "                elif nucleotide != ref_codon:\n",
    "                    if translator(nucleotide) == translator(ref_codon):\n",
    "                        synonymous_difference = 1\n",
    "                        synonymous_differences_list.append(synonymous_difference)\n",
    "                    \n",
    "                    elif translator(nucleotide) != translator(ref_codon):\n",
    "                        non_synonymous_difference = 1\n",
    "                        non_synonymous_differences_list.append(non_synonymous_difference)\n",
    "\n",
    "mean_number_of_synonymous_differences = statistics.mean(int(synonymous_differences_list))\n",
    "mean_number_of_non_synonymous_differences = statistics.mean(int(non_synonymous_differences_list))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estimating the proportion of synonymous and non-synonymous differences\n",
    "\n",
    "We can then therefore, estimate the proportion of synonymous ($p_s$) and non-synonymous($p_n$) differences by the following equations:\n",
    "$$\n",
    "\\begin{align}\n",
    "p_s = S_d / S \\\\\n",
    "p_n = N_d/N\\\\\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "To estimate the number of synonymous substitution ($d_s$) and non-synonymous substitutions ($d_N$) per site, the following formula developed by _Jukes and Cantor (1969)_ is used:\n",
    "\n",
    "$$\n",
    "\\begin{align}\n",
    "d = -\\frac{3}{4}log_e(1-\\frac{4}{3}p) \\\\\n",
    "\\text{where p is either }p_S \\text{ or } p_N\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "THis method gives only approximate estimates of $d_S$ and $d_N$, and is very accurate for more similar sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Calculation of p\n",
    "def proportion_of_synonymous_diff(mean_number_of_synonymous_differences, mean_number_of_synonymous_substitution):\n",
    "    proportion_of_synonymous_diff_val = mean_number_of_synonymous_differences / mean_number_of_synonymous_substitution\n",
    "    return proportion_of_synonymous_diff_val\n",
    "\n",
    "def proportion_of_non_synonymous_diff(mean_number_of_non_synonymous_differences, mean_number_non_synonymous_substitution):\n",
    "    proportion_of_non_synonymous_diff_val = mean_number_of_non_synonymous_differences / mean_number_non_synonymous_substitution\n",
    "    return proportion_of_non_synonymous_diff_val\n",
    "\n",
    "# Calculation of d\n",
    "import math\n",
    "def number_of_synonymous_substitution_per_site(proportion_of_synonymous_diff):\n",
    "    jukes_cantor_synonymous = 1 - ((4/3)*proportion_of_synonymous_diff)\n",
    "    number_of_synonymous_substitution_per_site_val = (-3/4)*(math.log(jukes_cantor_synonymous))\n",
    "    return number_of_synonymous_substitution_per_site_val\n",
    "\n",
    "\n",
    "def number_of_non_synonymous_substitution_per_site(proportion_of_non_synonymous_diff):\n",
    "    jukes_cantor_non_synonymous = 1 - ((4/3)*proportion_of_non_synonymous_diff)\n",
    "    number_of_non_synonymous_substitution_per_site_val = (-3/4)*(math.log(jukes_cantor_non_synonymous))\n",
    "    return number_of_non_synonymous_substitution_per_site_val"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Substitution rate per site ($\\omega$)\n",
    "\n",
    "Substitution rate per site,($\\omega_j$) can be calculated by a ration of non-synonymous substitution  : synonymous substitution\n",
    "\n",
    "\n",
    "$$\n",
    "\\omega_j = \\frac{number\\ of\\ synonymous\\ substitution\\ per\\ site}{number\\ of\\ non-synonymous\\ substitution\\ per\\ site}\n",
    "$$\n",
    "\n",
    "where $\\omega_j$ = substitution rate per site\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "omega_j = number_of_synonymous_substitution_per_site_val/number_of_non_synonymous_substitution_per_site_val"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2 (tags/v3.9.2:1a79785, Feb 19 2021, 13:44:55) [MSC v.1928 64 bit (AMD64)]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "6fff98fc3b3d81bd655c2cc48858186e4d9e2db7b515bf1c3221888f12a62f87"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
