{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# **S**ubstitution **I**nferences  using **M**ath and **B**ayesian **A**pproximation (**SIMBA**)\n",
    "\n",
    "SIMBA is a modification of the FUBAR analysis where instead of using maximum likelihood, we implemented a mathematical approach based on Nei and Gojobori, (1986), way of calculating number of synonymous and non-synonymous substitution. Markov Chain Monte Carlo (MCMC) Bayesian Inference is then used to estimate the best possible number of synonymous and non-synonymous rate to end up with calculating substitution rates. \n",
    "\n",
    "It is worthnoting that different approaches usually reach close values especially when working with similar codons. The SIMBA approach like most evolutionary approaches before it treats each codon as an independent unit undergoing selection without effects of neighbouring codons or homologous codons in other sequences. When working with too divergent sequences the obtained values might be biased, however because this is an approach to be useful when working with Spike genes of SARS CoV 2, this shall not be the case. \n",
    "\n",
    "SIMBA approach when looked into carefully, it can be seen that it follows the same routine as Fast Unconstrained Bayesian AppRoximation (FUBAR), only that the computer expensiveness of finding maximum likelihoods first though useful for a generalised tool cannot be useful for a tool with a specific similar sequences to analyse. Take Simba as a lightweight version of FUBAR specific for the Spike gene of SARS-CoV-2. \n",
    "\n",
    "A number of approaches used to ensure faster inferences include:\n",
    "1. Using a mathematical approach which is shorter than maximum likelihood apporach in FUBAR\n",
    "2. Assumption codons are independent evolutionary units.\n",
    "3. Exclusion of weighting since its going to be used on very similar amino acids.\n",
    "4. MCMC sampler to be based on pymc3."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Introduction\n",
    "\n",
    "Synonymous substitutions maybe used as a molecular clock for dating the evolutionary time of closely related species.\n",
    " \n",
    " <p align ='center'>\n",
    " <img\n",
    "    src = 'DNA triplet code.png'\n",
    ">\n",
    "</p>\n",
    "\n",
    " \n",
    "Genetic code table indicates that all substitutions at the second nucleotide positions of codons result in amino acid whereas a fraction of the nucleotide changes at the first and third positions are synonymous.\n",
    "\n",
    "Under the assumption of equal nucleoties frequencies and random substitution, this fraction is ~5 % for the first position and ~72 % for the thrid position.\n",
    "\n",
    "> $f_i$ = fraction of synoymous changes at the _i_th position of a given codon (i = 1,2,3)\n",
    "\n",
    "> _s_ = number of synonymous sites\n",
    "\n",
    "> n = number of non-synoymous sites.\n",
    "\n",
    "The _n_ and _s_ for this codon are then give by:\n",
    "\n",
    "$$\n",
    "\\begin{align}\n",
    "s = \\sum_{i=1}^{3} f_i \\  \\text{\\ and\\  n=(3-s)}\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "> using __Leu__ as an example,\n",
    "\n",
    "$f_1 = \\frac{1}{3}  \\text{A} \\to \\text{G}$\n",
    "\n",
    "using genetic code table, there is 1 in 3 chances that a change is from T $\\to$ C.\n",
    "\n",
    "$f_2$ = 0,   \n",
    "\n",
    "$f_3 =  \\frac{1}{3}(A \\to G)$ thus, \n",
    "\n",
    "$ S = \\frac{1}{3} + 0 + \\frac{1}{3} = \\frac{2}{3}$\n",
    "\n",
    "$ n = \\frac{2}{3} + \\frac{3}{3} + \\frac{2}{3} = \\frac{7}{3}$\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Spike Gene Cutter\n",
    "\n",
    "So far there is no simple process of cutting and cleaning Spike gene sequences from SARS-CoV-2 whole genome sequences. However, the user is supposed to use alignment of choice after the spike gene has been extracted. User can cut their genes using this block of code."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def spike_gene_cutter_cleaner(input_file, file_format):\n",
    "\n",
    "    # from Bio import SeqIO    # We use Biopython's SeqIO parser to load our sequences\n",
    "    import sys\n",
    "\n",
    "    # The  next section loads the SARS-CoV-2 sequences from file.\n",
    "\n",
    "    input_file = input(\"Enter full 'input' filename and filetype eg input_file.fasta\")\n",
    "    file_format = input(\"Enter 'file format' eg fasta\")\n",
    "\n",
    "    # Extraction of the approximate S gene\n",
    "    spike_genes = []        # All extracted spike genes will be stored in this list\n",
    "\n",
    "    for seq_record in SeqIO.parse(input_file, file_format):\n",
    "        spike_cut = seq_record[21500:25500]\n",
    "        spike_genes.append(spike_cut)\n",
    "    \n",
    "\n",
    "    spike_file = SeqIO.write(spike_genes, 'Spike_gene_file.fasta','fasta')\n",
    "\n",
    "    # Next part removes spike genes with a number of Unidentified nucleotides more than 2\n",
    "\n",
    "    clean_spike_genes = [] # This list will store cleaned spike genes\n",
    "\n",
    "    for spike_record in SeqIO.parse('Spike_gene_file.fasta','fasta'):\n",
    "        # first its wiser to use a string object of the sequence for counting additionally all letters should be ensured to be in upper case\n",
    "\n",
    "        string_spike = (str(spike_record.seq)).upper()\n",
    "\n",
    "        if string_spike.count('N') > 2:              #Unidentified nucleotides more than 2\n",
    "            if spike_record.seq not in clean_spike_genes:  # Removing repeat sequences\n",
    "                clean_spike_genes.append(spike_record)      #spike record is appended instead of string_spike so it does not lose Sequence object attributes\n",
    "\n",
    "\n",
    "print('Please check current folder for the Spike gene sequences which were extracted')\n",
    "\n",
    "\n",
    "# In case the user prefers to see extracted sequences\n",
    "print_sample = input('Do you want to view snips of extracted sequences: Y/N ')\n",
    "\n",
    "if print_sample.upper() == 'Y':\n",
    "    for spike in SeqIO.parse('Spike_gene_file.fasta','fasta'):\n",
    "            print(f'Length of approximate spike_gene = {len(spike)}')\n",
    "            print(f'Representative spike sequence = {repr(spike.seq)}')\n",
    "            print(f'Sequence id = {spike.id}')\n",
    "\n",
    "elif print_sample.upper() == 'N':\n",
    "        pass\n",
    "\n",
    "else:\n",
    "    print(\"Please enter 'Y' for 'Yes' and 'N' for 'No'\")\n",
    "\n",
    "# User can use alignment tool of choice and it is recommended to align against the GISAID Ref Seq for more accurate\n",
    "# downstream processes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The next step is to define the codon table which will be used to compare any substitutions to the GISAID reference sequence."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "\n",
    "ref_seq = SeqIO.read('GISAID Ref Seq.fasta', 'fasta')     # Reading the GISAID reference sequence into memory\n",
    "\n",
    "aligned_sequences = ()\n",
    "def triplet_code_table(triplet_code):\n",
    "\n",
    "    '''Triplet codes that code for specific amino acids based on the DNA triplet code table'''\n",
    "\n",
    "    for triplet_code in aligned_sequences:\n",
    "        # Gene sequence is a fasta file from a codon-aligned sequences, loaded using AlignIO of Biopython.\n",
    "        # Reminder that all loaded DNA sequences must be in upper case and in string format\n",
    "        if triplet_code == 'TTT' or 'TTC':\n",
    "            amino_acid = 'F'\n",
    "\n",
    "        elif triplet_code == 'TTA' or 'TTG' or 'CTT' or 'CTC'or 'CTA' or 'CTG':\n",
    "            amino_acid = 'L'\n",
    "\n",
    "        elif triplet_code == 'ATT' or 'ATC' or 'ATA':\n",
    "            amino_acid = 'I'\n",
    "\n",
    "        elif triplet_code == 'ATG':\n",
    "            amino_acid = 'M'\n",
    "\n",
    "        elif triplet_code == 'GTT' or 'GTC' or 'GTA' or 'GTG':\n",
    "            amino_acid = 'V'\n",
    "\n",
    "        elif triplet_code == 'TCT' or 'TCC' or 'TCA' or 'TCG' or 'AGT' or 'AGC':\n",
    "            amino_acid = 'S'\n",
    "\n",
    "        elif triplet_code == 'CCT' or 'CCC' or 'CCA' or 'CCG':\n",
    "            amino_acid = 'P'\n",
    "\n",
    "        elif triplet_code == 'ACT' or 'ACC' or 'ACA' or 'ACG':\n",
    "            amino_acid = 'T'\n",
    "\n",
    "        elif triplet_code == 'GCT' or 'GCC' or 'GCA' or 'GCG':\n",
    "            amino_acid = 'A'\n",
    "\n",
    "        elif triplet_code == 'TAT' or 'TAC':\n",
    "            amino_acid = 'Y'\n",
    "        \n",
    "        elif triplet_code == 'TAA' or 'TAG' or 'TGA':\n",
    "            amino_acid = '*'\n",
    "\n",
    "        elif triplet_code == 'CAT' or 'CAC':\n",
    "            amino_acid = 'H'\n",
    "\n",
    "        elif triplet_code == 'CAA' or 'CAG':\n",
    "            amino_acid = 'Q'\n",
    "\n",
    "        elif triplet_code == 'AAA' or 'AAG':\n",
    "            amino_acid = 'K'\n",
    "\n",
    "        elif triplet_code == 'GAA' or 'GAG':\n",
    "            amino_acid = 'E'\n",
    "\n",
    "        elif triplet_code == 'TGT' or 'TGC':\n",
    "            amino_acid = 'C'\n",
    "\n",
    "        elif triplet_code == 'TGG':\n",
    "            amino_acid = 'W'\n",
    "\n",
    "        elif triplet_code == 'CGT' or 'CGC' or 'CGA' or 'CGG' or 'AGA' or 'AGG':\n",
    "            amino_acid = 'R'\n",
    "\n",
    "        elif triplet_code == 'GGT' or 'GGC' or 'GGA' or 'GGG':\n",
    "            amino_acid = 'G'\n",
    "        \n",
    "        elif 'N' is in triplet_code:\n",
    "            amino_acid = '?'\n",
    "\n",
    "        else:\n",
    "            print('Please check if all bases are A, C, T or G')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that we have defined the codon table, it is time to calculate base substitution frequenciesin one codon."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def sum_of_codon_synonymous_substitutions(triplet_code, lower=0, upper=3):\n",
    "    for i in range(lower=0, upper=3):      #Lower represents the lower bound in our sum and upper is the upper bound. For python index, the first value would be 0\n",
    "        \n",
    "        if i = 0:                      \n",
    "            if base != ref_seq:            \n",
    "                fraction_of_change1 = 0.05            #Fraction of change denotes the chances of a codon for coding another amino acid when there is a mutation eg \n",
    "            else:\n",
    "                fraction_of_change1= 0                                    #changing the first codon nucleotide results in 5% chance of change in amino acid coded.                     \n",
    "        \n",
    "        elif i = 1:\n",
    "            if base != ref_seq:\n",
    "                fraction_of_change2 = 1\n",
    "            else:\n",
    "                fraction_of_change2 = 0\n",
    "\n",
    "        elif i = 2:\n",
    "            if base != ref_seq:\n",
    "                fraction_of_change3 = 0.72\n",
    "            else:\n",
    "                fraction_of_change3 = 0\n",
    "\n",
    "        sum_of_synonymous_fractions = sum(fraction_of_change1 + fraction_of_change2 + fraction_of_change3)              #gives use the sum of fraction of synonymous changes at ith position.\n",
    "        \n",
    "        return sum_of_synonymous_fractions\n",
    "\n",
    "\n",
    "def sum_of_non_synonymous_substitutions(sum_of_synonymous_fractions):\n",
    "    sum_of_non_synonymous_fractions = 3 - sum_of_synonymous_fractions   # to find sum of non-synonymous fraction we subtract the sum of \n",
    "    return sum_of_non_synonymous_fractions                               #synonymous fraction from 3\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Increasing number of codons\n",
    "\n",
    "For a DNA sequence of __r__ codons, the total number of synonymous and non-synonymous sites is therefore given by:\n",
    "\n",
    "$$ \n",
    "\\begin{align*}\n",
    "S = \\sum_{j=1}^r S_i \\     \\text{and\\   n = (3r -S)} \\\\\n",
    "\\text{where} \\  s_i = \\text{value of s for the i-th codon} \\\\\n",
    "\\text{j = position number of codon j in DNA sequence with r codons}\n",
    "\\end{align*}\n",
    "$$\n",
    "\n",
    "\n",
    "When two sequences are compared, the averages of __S__ and __N__ are used."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Computing nucleotide differences between a pair of homologous sequences\n",
    "\n",
    "To compute the number of synonymous and non synonymous nucleotide differences between a pair of homologous sequences, we compare the two sequences codon, we compare the two sequences codon by codon and count the number of synoymous and non-synonymous nucleotide differences  for each pair of codon compared.\n",
    "\n",
    "When there is only one nucleotide difference, we can immediately decide whether the substitution is synonymous or non-synonymous.\n",
    "> eg, if the codon pairs compared are __GTT__ (_Val_) and __GTA__ (_Val_), there is one synonymous difference.\n",
    "\n",
    "We denote $S_d$ and $n_d$ the number of synonymous and non-synoynymous differences per codon, respecitively. In the prent case, $S_d = 1$ and $n_d = 0$.\n",
    "\n",
    "For example, in the comparison of __TTT__ and __GTA__, the two pathways are as follows:\n",
    "\n",
    "> _Pathway I_:\n",
    "> > __TTT__ (_Phe_) $\\to$ __GTT__ (_Val_) $\\to$ __GTA__ (_Val_)\n",
    "\n",
    "> _Pathway II_:\n",
    "> > __TTT__ (_Phe_) $\\to$ __TTA__ (_Leu_) $\\to$ __GTA__ (_Val_)\n",
    "\n",
    "Pathway I involves one synonymous and one non-synonymous substitution whereas Pathway II involves two non-synonymous  substitutions.\n",
    "\n",
    "We assume that pathway I and II occur with equal probability.\n",
    "\n",
    "The $s_d$ and &n_d$ then become 0.5 and 1.5 repsectively.\n",
    "\n",
    "When there are three nucleotide differences between the codons compared, there are six different possible pathwaysbetween the codons [3(x-1)] and in each pathway there are 3 mutation steps [x] where x = nucleotide differences.\n",
    "\n",
    "It is now clear that the total number of synonymous and non-synonymous differences can be obtained by summing up these values over all codons i.e.\n",
    "$$\n",
    "\\begin{align*}\n",
    "S_d = \\sum_{j=1}^{r} s_{dj} \\\\\n",
    "N_d = \\sum_{j=1}^{r} n_{dj} \\\\\n",
    "\\text{where} \\ s_{dj} \\text{\\  and } n_{dj} \\text{\\  are } s_d \\text{\\ and } n_d \\text{ for the j-th codon respectively} \\\\\n",
    "\\text{and r is the number of codons compared}\n",
    "\\end{align*}\n",
    "$$\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Estimating the proportion of synonymous and non-synonymous differences\n",
    "\n",
    "We can then therefore, estimate the proportion of synonymous ($p_s$) and non-synonymous($p_n$) differences by the following equations:\n",
    "$$\n",
    "\\begin{align}\n",
    "p_s = S_d / S \\\\\n",
    "p_n = N_d/N\\\\\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "To estimate the number of synonymous substitution ($d_s$) and non-synonymous substitutions ($d_N$) per site, the following formula developed by _Jukes and Cantor (1969)_ is used:\n",
    "\n",
    "$$\n",
    "\\begin{align}\n",
    "d = -\\frac{3}{4}log_e(1-\\frac{4}{3}p) \\\\\n",
    "\\text{where p is either }p_S \\text{ or } p_N\n",
    "\\end{align}\n",
    "$$\n",
    "\n",
    "THis method gives only approximate estimates of $d_S$ and $d_N$, and is very accurate for more similar sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 ('base')",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12 (main, Apr  4 2022, 05:22:27) [MSC v.1916 64 bit (AMD64)]"
  },
  "orig_nbformat": 4,
  "vscode": {
   "interpreter": {
    "hash": "57bc2b6ce032b5f0e93daa91901b7ea38a856826ef43aa9e95b6d3999f5310df"
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
